File: checkout_git_repositories.js
var Fiber = require("fibers");
var Future = require("fibers/future"),
  wait = Future.wait;
var fs = require('fs');

if (process.argv.length != 4) {
  console.log('Usage: node checkout_git_repositories <git server> <directory>');
  process.exit(1);
}

var gitServer = process.argv[2];
var repositoriesDirectory = process.argv[3];

if (!fs.existsSync(repositoriesDirectory)) {
  console.warn('Directory does not exist and should be created first: ' + repositoriesDirectory);
  process.exit(2);
}

checkoutRepository('gitolite-admin', true);
checkoutRepositories();


function checkoutRepositories() {
  var confsDirectory = repositoriesDirectory + '/gitolite-admin/conf/subs';
  console.log('Checking out repositories: ' + confsDirectory);

  fs.readdir(confsDirectory, function(err, files) {
    if (err) {
      console.log('Error: ' + err);
      process.exit(1);
    }
    files.forEach(function(confFile) {
      parseFile(confsDirectory + '/' + confFile);
    });
  });
}

function parseFile(confFile) {
  console.log('Processing file: ' + confFile);

  var lineByLineReader = require('line-by-line');
  var lr = new lineByLineReader(confFile);

  lr.on('error', function(err) {
    console.log('Error reading file ' + confFile + ': ' + err);
  });

  lr.on('line', function(line) {
    if (line.indexOf('repo ') == 0) {
      var repository = line.substring(5);
      checkoutRepository(repository, false);
    }
  });
}

function checkoutRepository(repository, pullIfExists) {
  console.log('Checking out repository: ' + repository);

  var directory = repositoriesDirectory + '/' + repository;
  var exists = fs.existsSync(directory);
  if (exists) {
    if (pullIfExists) {
      gitPull(repository, directory);
    }
  } else {
    gitClone(repository, directory);
  }
}

function gitPull(repository, directory) {
  console.log('Git pull ' + repository + ' in ' + directory);

  var options = {
    cwd: directory
  };
  git(['pull'], options);
}

function gitClone(repository, directory) {
  console.log('Git clone ' + repository + ' into ' + directory);

  git(['clone', 'git@' + gitServer + ':' + repository, directory]);
}

function git(args, options) {
  var future = new Future;

  var child_process = require('child_process');
  var git = child_process.spawn('echo', args, options);

  git.stdout.on('data', function(data) {
    //console.log('stdout: ' + data);
  });

  git.stderr.on('data', function(data) {
    //console.log('stderr: ' + data);
  });

  git.on('exit', function(code) {
    if (code != 0) {
      console.log('Child process exited with exit code ' + code);
    }
    future.
    return ();
  });

  future.resolve(function() {});
}

File: checkout_git_repositories.rb
#!/usr/bin/env ruby

if ARGV.length != 2
  puts 'Usage: checkout_git_repositories.rb <git server> <directory>'
  exit 1
end

$git_server = ARGV[0];
$repositories_directory = ARGV[1];

if !File.directory?($repositories_directory)
  puts "Directory does not exist and should be created first: #{repositories_directory}"
  exit 2
end


def checkout_repositories()
  confsDirectory = $repositories_directory + '/gitolite-admin/conf/subs'
  puts "Checking out repositories: #{confsDirectory}"

  Dir.foreach(confsDirectory) do |confFile|
    next if confFile == '.' or confFile == '..'
    parse_file("#{confsDirectory}/#{confFile}")
  end
end

def parse_file(confFile)
  puts "Processing file: #{confFile}"

  f = File.open(confFile)
  f.each do |line|
    next if !line.start_with?('repo ')
    repository = line[5..-1]
    checkout_repository(repository, false)
  end
end

def checkout_repository(repository, pull_if_exists)
  puts "Checking out repository: #{repository}"

  directory = $repositories_directory + '/' + repository;
  exists = File.directory?(directory);
  if exists
    if pull_if_exists
      git_pull(repository, directory)
    end
  else
    git_clone(repository, directory)
  end
end

def git_pull(repository, directory)
  puts "Git pull #{repository} in #{directory}"
  `cd #{directory} && git pull`
end

def git_clone(repository, directory)
  puts "Git clone #{repository} into #{directory}"
  `git clone git@#{$git_server}:#{repository} #{directory}`
end


checkout_repository('gitolite-admin', true)
checkout_repositories()